components/playlist/browser/playlist_service.cc

```diff
@@ -0,0 +1,575 @@
+}
+
+void PlaylistService::OnThumbnailDownloaded(const std::string& id,
+                                            SimpleURLLoaderList::iterator it,
+                                            base::FilePath path) {
+  // When delete all is requested during the thumbnail downloading, we should
+  // just return. |url_loaders_| is cleared.
+  if (url_loaders_.empty())
+    return;
+
+  url_loaders_.erase(it);
+
+  if (path.empty()) {
+    VLOG(2) << __func__ << ": thumbnail fetching failed for " << id;
+    NotifyPlaylistChanged(
+        {PlaylistChangeParams::ChangeType::CHANGE_TYPE_THUMBNAIL_FAILED, id});
+    return;
+  }
+
+  const std::string thumbnail_path =
+#if defined(OS_WIN)
+      base::UTF16ToUTF8(path.value());
+#else
+      path.value();
+#endif
+
+  const base::Value* value = prefs_->Get(kPlaylistItems)->FindDictKey(id);
+  DCHECK(value);
+  if (value) {
+    base::Value copied_value = value->Clone();
+    copied_value.SetStringKey(kPlaylistThumbnailPathKey, thumbnail_path);
+    UpdatePlaylistValue(id, std::move(copied_value));
+    NotifyPlaylistChanged(
+        {PlaylistChangeParams::ChangeType::CHANGE_TYPE_THUMBNAIL_READY, id});
+  }
+}
+
+base::Value PlaylistService::GetAllPlaylistItems() {
+  base::Value playlist(base::Value::Type::LIST);
+  for (const auto& it : prefs_->Get(kPlaylistItems)->DictItems()) {
+    base::Value item = it.second.Clone();
+    item.RemoveKey(kPlaylistCreateParamsKey);
+    playlist.Append(std::move(item));
+  }
+
+  return playlist;
+}
+
+base::Value PlaylistService::GetPlaylistItem(const std::string& id) {
+  const base::Value* item_value_ptr =
+      prefs_->Get(kPlaylistItems)->FindDictKey(id);
+  if (item_value_ptr) {
+    base::Value item = item_value_ptr->Clone();
+    item.RemoveKey(kPlaylistCreateParamsKey);
+    return item;
+  }
+  return {};
+}
+
+void PlaylistService::RecoverPlaylistItem(const std::string& id) {
+  const base::Value* playlist_info =
+      prefs_->Get(kPlaylistItems)->FindDictKey(id);
+  if (!playlist_info) {
+    LOG(ERROR) << __func__ << ": Invalid playlist id for recover: " << id;
+    return;
+  }
+
+  VLOG(2) << __func__ << ": This is in recovering";
+
+  const std::string* thumbnail_path_str =
+      playlist_info->FindStringPath(kPlaylistThumbnailPathKey);
+  const bool has_thumbnail =
+      !!thumbnail_path_str && !thumbnail_path_str->empty();
+  if (!has_thumbnail)
+    DownloadThumbnail(id);
+
+  base::Optional<bool> partial_ready =
+      playlist_info->FindBoolPath(kPlaylistPartialReadyKey);
+  const std::string* video_media_file_path =
+      playlist_info->FindStringPath(kPlaylistVideoMediaFilePathKey);
+  const std::string* audio_media_file_path =
+      playlist_info->FindStringPath(kPlaylistAudioMediaFilePathKey);
+  // Only try to regenerate if partial ready or there is no media file.
+  if (!video_media_file_path || video_media_file_path->empty() ||
+      !audio_media_file_path || audio_media_file_path->empty() ||
+      *partial_ready) {
```

Comment:
this will DCHECK if `Optional` has no value. Should we check `has_value()`?

Reply:
No, `kPlaylistPartialReadyKey` value will alway be set.